File: DreamTeam/src/main/java/team/project/redboost/entities/Task.java
package team.project.redboost.entities;

import jakarta.persistence.*;
import lombok.Data;
import com.fasterxml.jackson.annotation.JsonBackReference;
import com.fasterxml.jackson.annotation.JsonManagedReference;
import lombok.ToString;

import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

@Data
@Entity
@Table(name = "tasks")
public class Task {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long taskId;

    @Column(nullable = false)
    private String title;

    private int xpPoint;

    @Column(columnDefinition = "TEXT")
    private String description;

    @Column(name = "assignee_id")
    private Long assigneeId;

    @Column(name = "start_date")
    private LocalDate startDate;

    @Column(name = "end_date")
    private LocalDate endDate;

    @Enumerated(EnumType.STRING)
    private Priority priority;

    @Enumerated(EnumType.STRING)
    private Status status;

    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "phase_id", nullable = false)
    @JsonBackReference("phaseTasks")
    @ToString.Exclude
    private Phase phase;

    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "task_category_id", nullable = false)
    @JsonBackReference("categoryTasks")
    @ToString.Exclude
    private TaskCategory taskCategory;

    @Column(name = "created_at", updatable = false)
    private LocalDateTime createdAt;

    @Column(name = "updated_at")
    private LocalDateTime updatedAt;

    @Column
    private String attachment; // Stores Google Drive file ID

    @Transient
    private Long taskCategoryId;

    @OneToMany(mappedBy = "task", cascade = CascadeType.ALL, orphanRemoval = true, fetch = FetchType.LAZY)
    @JsonManagedReference("taskSubTasks")
    @ToString.Exclude
    private List<SubTask> subTasks = new ArrayList<>();

    @OneToMany(mappedBy = "task", cascade = CascadeType.ALL, orphanRemoval = true, fetch = FetchType.LAZY)
    @JsonManagedReference("taskComments")
    @ToString.Exclude
    private List<Comment> comments = new ArrayList<>();

    public enum Priority {
        LOW, MEDIUM, HIGH
    }

    public enum Status {
        TO_DO, IN_PROGRESS, DONE, VALIDATED
    }

    @PrePersist
    protected void onCreate() {
        createdAt = LocalDateTime.now();
        updatedAt = LocalDateTime.now();
        setTaskCategoryId();
    }

    @PreUpdate
    protected void onUpdate() {
        updatedAt = LocalDateTime.now();
        setTaskCategoryId();
    }

    @PostLoad
    protected void setTaskCategoryId() {
        if (taskCategory != null) {
            this.taskCategoryId = taskCategory.getId();
        }
    }

    public void addSubTask(SubTask subTask) {
        subTasks.add(subTask);
        subTask.setTask(this);
    }

    public void addComment(Comment comment) {
        comments.add(comment);
        comment.setTask(this);
    }
}

File: Redboost_front/src/app/models/task.ts
import { Phase } from './phase';

export interface TaskCategory {
    id?: number;
    name: string;
}

export enum Priority {
    LOW = 'LOW',
    MEDIUM = 'MEDIUM',
    HIGH = 'HIGH'
}

export const PriorityTranslation: Record<Priority, string> = {
    [Priority.LOW]: 'Faible',
    [Priority.MEDIUM]: 'Moyenne',
    [Priority.HIGH]: 'Élevée'
};

export enum Status {
    TO_DO = 'TO_DO',
    IN_PROGRESS = 'IN_PROGRESS',
    DONE = 'DONE',
    VALIDATED = 'VALIDATED'
}

export const StatusTranslation: Record<Status, string> = {
    [Status.TO_DO]: 'À faire',
    [Status.IN_PROGRESS]: 'En cours',
    [Status.DONE]: 'Terminé',
    [Status.VALIDATED]: 'Validé'
};

export interface Attachment {
    name: string;
    url?: string;
    fileId?: string;
}

export interface SubTask {
    subTaskId?: number;
    title: string;
    description?: string;
}

export interface Comment {
    commentId?: number;
    content: string;
    profilePictureUrl?: string;
    firstName: string;
    lastName: string;
    userId: number;
    createdAt?: string;
}

export interface Task {
    taskId?: number;
    title: string;
    xpPoint: number;
    description?: string;
    assigneeId?: number;
    startDate?: string;
    endDate?: string;
    priority: Priority;
    status: Status;
    phase: Phase;
    taskCategory: TaskCategory;
    createdAt?: string;
    updatedAt?: string;
    attachment: Attachment;
    taskCategoryId?: number;
    subTasks?: SubTask[];
    comments?: Comment[];
}
